features/eolearn/features/temporal_features.py

```diff
@@ -243,3 +245,369 @@ def execute(self, eopatch):
+            pos = np.zeros(len(derivatives) + 2)
+            pos[1:len(derivatives)+1] = derivatives >= 0
+
+            pos_der_int, pos_der_len, pos_der_rate, (start, _) = \
+                self.derivative_features(pos, valid_dates, data_curve, self.base_surface_min)
+
+            data_pos_surf[ih, iw] = pos_der_int
+            data_pos_len[ih, iw] = pos_der_len
+            data_pos_rate[ih, iw] = pos_der_rate
+
+            neg = np.zeros(len(derivatives) + 2)
+            neg[1:len(derivatives)+1] = derivatives <= 0
+
+            neg_der_int, neg_der_len, neg_der_rate, (_, end) = \
+                self.derivative_features(neg, valid_dates, data_curve, self.base_surface_min)
+
+            data_neg_surf[ih, iw] = neg_der_int
+            data_neg_len[ih, iw] = neg_der_len
+            data_neg_rate[ih, iw] = neg_der_rate
+
+            if self.ndvi_feature_name:
+                data_pos_tr[ih, iw] = \
+                    np.any(eopatch[self.ndvi_feature_name[0]][self.ndvi_feature_name[1]][:start+1, ih, iw, 0] <=
+                           self.ndvi_barren_soil_cutoff)
+                data_neg_tr[ih, iw] = \
+                    np.any(eopatch[self.ndvi_feature_name[0]][self.ndvi_feature_name[1]][end:, ih, iw, 0] <=
+                           self.ndvi_barren_soil_cutoff)
+
+        eopatch.data_timeless[self.max_val_feature] = data_max_val[..., np.newaxis]
+        eopatch.data_timeless[self.min_val_feature] = data_min_val[..., np.newaxis]
+        eopatch.data_timeless[self.mean_val_feature] = data_mean_val[..., np.newaxis]
+        eopatch.data_timeless[self.sd_val_feature] = data_sd_val[..., np.newaxis]
+
+        eopatch.data_timeless[self.diff_max_feature] = data_diff_max[..., np.newaxis]
+        eopatch.data_timeless[self.diff_min_feature] = data_diff_min[..., np.newaxis]
+        eopatch.data_timeless[self.diff_diff_feature] = (data_diff_max - data_diff_min)[..., np.newaxis]
+
+        eopatch.data_timeless[self.max_mean_feature] = data_max_mean[..., np.newaxis]
+        eopatch.data_timeless[self.max_mean_len_feature] = data_max_mean_len[..., np.newaxis]
+        eopatch.data_timeless[self.max_mean_surf_feature] = data_max_mean_surf[..., np.newaxis]
+
+        eopatch.data_timeless[self.pos_len_feature] = data_pos_len[..., np.newaxis]
+        eopatch.data_timeless[self.pos_surf_feature] = data_pos_surf[..., np.newaxis]
+        eopatch.data_timeless[self.pos_rate_feature] = data_pos_rate[..., np.newaxis]
+        eopatch.data_timeless[self.pos_transition_feature] = data_pos_tr[..., np.newaxis]
+
+        eopatch.data_timeless[self.neg_len_feature] = data_neg_len[..., np.newaxis]
+        eopatch.data_timeless[self.neg_surf_feature] = data_neg_surf[..., np.newaxis]
+        eopatch.data_timeless[self.neg_rate_feature] = data_neg_rate[..., np.newaxis]
+        eopatch.data_timeless[self.neg_transition_feature] = data_neg_tr[..., np.newaxis]
+
+        return eopatch
+
+    def get_data(self, patch):
```

Comment:
I don't think this function belongs here, because it just serves your particular case of data arrangement. All tasks get an eopatch, process it and return an eopatch. I think that here you convert the data into a form that you need in your particular application. Perhaps this data re-arrangement should also happen there in the app instead of in eo-learn, which should be more general.